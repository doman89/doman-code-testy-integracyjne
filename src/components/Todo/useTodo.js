import { useState, useRef } from "react";

export function useTodo() {
  const [todos, setTodos] = useState([]);
  const inputRef = useRef(null);

  const addTodo = () => {
    if (!inputRef.current) {
      return;
    }

    const label = inputRef.current.value;

    setTodos((prev) =>
      prev.concat([{ id: Date.now(), label }]),
    );
    inputRef.current.value = "";
  };

  const removeTodo = ({ currentTarget }) => {
    const { id } = currentTarget.dataset;

    setTodos((prev) =>
      prev.filter((todo) => todo.id !== Number(id)),
    );
  };

  return { 
    addTodo, 
    inputRef, 
    removeTodo,
    todos 
  }
}