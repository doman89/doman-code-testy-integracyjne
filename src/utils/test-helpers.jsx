import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { cloneElement } from "react";

export function snapshotComponent(component) {
  const { asFragment, getByTestId } = render(
    cloneElement(component, {
      "data-testid": "test",
    }),
  );

  const renderedComponent = getByTestId("test");

  expect(renderedComponent).toBeDefined();
  expect(asFragment()).toMatchSnapshot();
}

export function renderComponent(component) {
  const user = userEvent.setup();

  render(component);

  return { screen, user };
}

export function mockDateNow(time) {
  jest.spyOn(global.Date, "now").mockImplementation(() => time);
}
