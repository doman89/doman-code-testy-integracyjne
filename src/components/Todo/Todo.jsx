import { useRef, useState } from "react";

import { Button } from "../Button";
import { Input } from "../Input";
import { List } from "../List";
import { ListElement } from "../ListElement";

import "./todo.css";
import { useTodo } from "./useTodo";

export function Todo({ "data-testid": dataTestId }) {
  const { 
    addTodo, 
    inputRef, 
    removeTodo,
    todos 
  } = useTodo();

  const rows = todos.map((element) => (
    <ListElement
      key={element.id}
      {...element}
      onRemove={removeTodo}
    />
  ));

  return (
    <div className="wrapper" data-testid={dataTestId}>
      <div className="row">
        <Input data-testid="input" ref={inputRef} />
        <Button
          data-testid="add-button"
          variant="add"
          onClick={addTodo}
        >
          +
        </Button>
      </div>
      <List elements={rows}>{rows}</List>
    </div>
  );
}
