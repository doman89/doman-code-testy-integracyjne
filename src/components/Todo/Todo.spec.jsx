import { mockDateNow, renderComponent, snapshotComponent } from "../../utils/test-helpers";

import { Todo } from "./Todo";
import * as useTodoMock from "./useTodo";

describe("<Todo />", () => {
  describe("integration", () => {
    it("should work as expected", async () => {
      mockDateNow(1);
      const { screen, user } = renderComponent(<Todo />);
      const button = screen.getByTestId("add-button");
      const input = screen.getByTestId("input");

      expect(screen.getByTestId("no-results")).toBeInTheDocument();

      await user.type(input, "Doman Code");
      await user.click(button);

      const label = await screen.findByTestId("todo-1");

      expect(label).toHaveTextContent("Doman Code");

      const removeButton = screen.getByTestId("remove-button-1");

      await user.click(removeButton);

      expect(screen.getByTestId("no-results")).toBeInTheDocument();
    });
  });

  describe("only UI testing", () => {
    beforeEach(() => {
      mockUseTodo();
    });

    it("should match snapshot without data", () => {
      snapshotComponent(<Todo />);
    });

    it("should match snapshot with data", () => {
      mockUseTodo({
        todos: [
          { id: "test", label: "test" },
          { id: "Doman", label: "Doman Code" },
        ]
      });
      snapshotComponent(<Todo />);
    });

    it("should call addTodo function when '+' button has been clicked", async () => {
      const addTodo = jest.fn();

      mockUseTodo({
        addTodo,
      });

      const { screen, user } = renderComponent(<Todo />);
      const button = screen.getByTestId("add-button");

      await user.click(button);

      expect(addTodo).toHaveBeenCalled();
    });

    it("should call removeTodo function when '-' button has been clicked", async () => {
      const removeTodo = jest.fn();

      mockUseTodo({
        todos: [{ id: 1 }],
        removeTodo,
      });

      const { screen, user } = renderComponent(<Todo />);
      const button = screen.getByTestId("remove-button-1");

      await user.click(button);

      expect(removeTodo).toHaveBeenCalled();
    });

    it("should have correct label base on data in todo array", () => {
      mockUseTodo({
        todos: [{ id: 1, label: "Doman Code" }],
      });

      const { screen } = renderComponent(<Todo />);
      const label = screen.getByTestId("todo-1");

      expect(label).toBeInTheDocument();
    });

    it("should have value in input after user interaction", async () => {
      const { screen, user } = renderComponent(<Todo />);
      const input = screen.getByTestId("input");

      await user.type(input, "Doman Code");

      expect(input).toHaveValue("Doman Code");
    });

    it("should show information about lack of data when array is empty", () => {
      const { screen } = renderComponent(<Todo />);
      const noResult = screen.getByTestId("no-results");

      expect(noResult).toBeInTheDocument();
    });

    it("should render all elements", () => {
      mockUseTodo({
        todos: [
          { id: 1 },
          { id: 2 },
          { id: 3 },
          { id: 4 },
          { id: 5 },
        ],
      });
      const { screen } = renderComponent(<Todo />);
      const list = screen.getByTestId("todo-list");

      expect(list.childElementCount).toBe(5);
    });
  });
});

const useTodoMockData = { 
  addTodo: jest.fn(), 
  inputRef: { current: null }, 
  removeTodo: jest.fn(),
  todos: [],
}

function mockUseTodo(mockData) {
  jest.spyOn(useTodoMock, "useTodo").mockImplementation(() => ({
    ...useTodoMockData,
    ...mockData,
  }));
}