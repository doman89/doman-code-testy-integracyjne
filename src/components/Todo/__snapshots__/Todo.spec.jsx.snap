// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo /> only UI testing should match snapshot with data 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="test"
  >
    <div
      class="row"
    >
      <input
        class="input"
        data-testid="input"
      />
      <button
        class="button button--green"
        data-testid="add-button"
      >
        +
      </button>
    </div>
    <ul
      class="list-wrapper"
      data-testid="todo-list"
    >
      <li
        class="list-element"
      >
        <span
          data-testid="todo-test"
        >
          test
        </span>
        <button
          class="button button--red"
          data-id="test"
          data-testid="remove-button-test"
        >
          -
        </button>
      </li>
      <li
        class="list-element"
      >
        <span
          data-testid="todo-Doman"
        >
          Doman Code
        </span>
        <button
          class="button button--red"
          data-id="Doman"
          data-testid="remove-button-Doman"
        >
          -
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<Todo /> only UI testing should match snapshot without data 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="test"
  >
    <div
      class="row"
    >
      <input
        class="input"
        data-testid="input"
      />
      <button
        class="button button--green"
        data-testid="add-button"
      >
        +
      </button>
    </div>
    <ul
      class="list-wrapper"
      data-testid="no-results"
    >
      Nie ma elementów do wyświetlenia
    </ul>
  </div>
</DocumentFragment>
`;
